C API: Burn `Optimize1qGatesDecomposition`.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

After oxidizing `Optimize1QGatesDecomposition` in #12959  it's time to make a combustion reaction that releases a carbon byproduct (our C API for it). The following commits aim to expose a standalone function that will give access to the rust counterpart.

This implementation adds the function ``qk_transpiler_standalone_optimize_1q_gates_decomposition`` which accepts a ``QkCircuit`` and an optional instance of ``QkTarget`` and calls the Rust counterpart.

### Details and comments

Closes #14453. ~~Built on top of #14668 (NEEDS REBASING AFTER IT MERGES).~~ REBASED!

raynelfss committed:
Initial: Introduce C-API Access to `Optimize1qGatesDecomposition`.
- The following commit adds initial access to a standalone C foreign function interface that allows users to run the `Optimize1qGatesDecomposition` transpiler pass. The feature as it stands requires the user to build a target prior to submitting the circuit; This might change in the future.
- Added 1 initial testing suite for the pass which converts a sequence of H gates into a different sequences depending on the provided target. More tests will be added with time.

raynelfss committed:
Fix: Add target identity test.

raynelfss committed:
Chore: Fix docstring and free some pointers.

raynelfss committed:
Add: Support for optional target for `Optimize1qGatesDecomposition` C API.
- Add ability of running the pass without passing a `Target`.
- Add proper formatting to unit tests with multiple tests inside.
- Add tests without target.

raynelfss committed:
Chore: Add release note.

raynelfss committed:
Chore: Fix docstring.

raynelfss committed:
Chore: Fix bug in release note

raynelfss commented:
This PR is currently blocked on #14777

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,110 @@
+/// Runs the Optimize1qGatesDecomposition pass in standalone mode on a circuit.
+///
+/// Optimize1qGatesDecomposition, as its name implies, optimizes chains of single-qubit
+/// gates by combining them into a single gate.
```
This pass doesn't combine gates into a single qubit gate though, does it? It only does so temporarily but outputs a sequence of gates. We probably should rephrase this accordingly (also the Python docs) 🙂

raynelfss replied:
Updated in 923a8bb19018f55875382f5db28bc4d4c6dfb3e0

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,110 @@
+/// qubit it operates on.
+///
+/// @param circuit A pointer to the ``QkCircuit`` object to transform.
+/// @param target A pointer to the ``QkTarget`` object.
```
From the code it looks like this is allowed to be a null pointer. What happens in this case? We should document that

raynelfss replied:
From snippet:
```rs
/// In the case a null pointer is provided and gate errors are unknown
/// the pass will choose the sequence with the least amount of gates,
/// and will support all basis gates on its Euler basis set.
```

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,110 @@
+        .expect("Error while converting the dag to a circuit.");
+
+    // Convert to pointer.
+    Box::into_raw(Box::new(dag_to_circuit))
```
I think we recently decided that the passes should run in-place, right?

raynelfss replied:
Addressed in cf16fa9aa81b08d6bab87ac4b25fe86d7c01d712

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,110 @@
+/// if ``target`` is not a valid pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpiler_standalone_optimize_1q_gates_decomposition(
```
I'm wondering if we should take this chance and rename it to something more expressive, like
```
qk_transpiler_standalone_optimize_1q_sequences
```
or something. Or do we want to keep a 1:1 naming scheme across Qiskit's APIs?

raynelfss replied:
Name finally updated in 923a8bb19018f55875382f5db28bc4d4c6dfb3e0

mtreinish commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,110 @@
+///     // Clean up
+///     qk_target_free(target_u1_u2_u3);
+///     qk_circuit_free(circuit);
+///     qk_circuit_free(circuit_result);
```
```suggestion
///     QkTarget *target = qk_target_new(1);
///
///     double u_errors[3] = {0., 1e-4, 1e-4};
///         QkTargetEntry *u_entry = qk_target_entry_new(QkGate_U);
///         uint32_t qargs[1] = {0,};
///         qk_target_entry_add_property(u_entry, qargs, 1, NAN, u_errors[idx]);
///         qk_target_add_instruction(target, u_entry);
///
///     // Build circuit
///     QkCircuit *circuit = qk_circuit_new(1, 0);
///     uint32_t qubits[1] = {0};
///     for (int iter = 0; iter < 3; iter++) {
///         qk_circuit_gate(circuit, QkGate_H, qubits, NULL);
///     }
///
///     // Run transpiler pass
///     QkCircuit *circuit_result =
///         qk_transpiler_standalone_optimize_1q_gates_decomposition(circuit, target);
///     
///     // Clean up
///     qk_target_free(target);
///     qk_circuit_free(circuit);
///     qk_circuit_free(circuit_result);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+
+    double *u_params[3] = {u1_params, u2_params, u3_params};
+    for (int idx = 0; idx < 3; idx++) {
+        QkTargetEntry *u_entry = qk_target_entry_new_fixed(u_gates[idx], u_params[idx]);
```
```suggestion
        QkTargetEntry *u_entry = qk_target_entry_new(u_gates[idx]);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+    double r_params[1] = {3.14};
+
+    for (int idx = 0; idx < 2; idx++) {
+        QkTargetEntry *r_entry = qk_target_entry_new_fixed(r_gates[idx], r_params);
```
```suggestion
        QkTargetEntry *r_entry = qk_target_entry_new(r_gates[idx]);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+    for (int idx = 0; idx < 2; idx++) {
+        QkTargetEntry *entry;
+        if (gates[idx] == QkGate_RZ) {
+            entry = qk_target_entry_new_fixed(gates[idx], rz_params);
```
```suggestion
            entry = qk_target_entry_new(gates[idx]);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+
+    double *gate_params[3] = {rz_params, rx_params, u_params};
+    for (int idx = 0; idx < 3; idx++) {
+        QkTargetEntry *u_entry = qk_target_entry_new_fixed(u_gates[idx], gate_params[idx]);
```
```suggestion
        QkTargetEntry *u_entry = qk_target_entry_new(u_gates[idx]);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+    for (int idx = 0; idx < 2; idx++) {
+        QkTargetEntry *entry;
+        if (gates[idx] == QkGate_Phase) {
+            entry = qk_target_entry_new_fixed(gates[idx], phase_params);
```
```suggestion
            entry = qk_target_entry_new(gates[idx]);
```

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+        qk_target_add_instruction(target_h_p, entry);
+    }
+    return target_h_p;
+}
```
This doesn't really offer much value this is not a 1q universal basis set so we won't ever have a decomposer for it.

mtreinish commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,378 @@
+    fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);
+
+    return num_failed;
+}
```
Can you move this to the bottom of the file?

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-optimize-1q-decomp

raynelfss committed:
Chore: Address review comments
- Make tha function call run in-place and modify the pointer to the `QkCircuit` in use.
- Remove `h_p_target` from tests.
- Use parametric gates in the target with generic parameter expressions instead of fixed ones.

raynelfss committed:
Update crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Chore: Fix docstring

raynelfss committed:
Chore: Fix release note

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,103 @@
+/// @param target A pointer to the ``QkTarget`` object or a null pointer.
+/// In the case a null pointer is provided the pass calculates the error
+/// of each instruction based on each one-qubit sequence's length, and
+/// will support all basis gates on its Euler basis set.
```
I'm not sure I fully understand the current formulation: if we don't have the target we can't calculate any errors right? Is the following suggestion maybe clearer (assuming that it is correct what I wrote 😄)?
```suggestion
/// In the case a null pointer is provided and gate errors are unknown
/// the pass will choose the sequence with the least amount of gates,
/// and will support all basis gates on its Euler basis set.
```

raynelfss replied:
Added to 19f55a61197d9d6ac5b188778e4a2dd98b020b1c

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,103 @@
+/// of each instruction based on each one-qubit sequence's length, and
+/// will support all basis gates on its Euler basis set.
+///
+/// @return The circuit after applying the optimizations.
```
We need to either remove this or make this return an exit code 🙂

raynelfss replied:
Oh right, I forgot to update the docstring. Should we have all the passes return exit codes then? I know we do modifications in place which frees us from returning extra structs.

raynelfss replied:
Addressed in 19f55a61197d9d6ac5b188778e4a2dd98b020b1c

Cryoris commented on crates/cext/src/transpiler/passes/optimize_1q_decomposition.rs:
```diff
@@ -0,0 +1,103 @@
+
+    // Convert the circuit to a DAG.
+    let mut circuit_as_dag = DAGCircuit::from_circuit_data(circuit, false, None, None, None, None)
+        .expect("Error while converting the circuit to a dag.");
```
It would be nice if we would have a set of `TranspilerError`s which we could return here instead of just panicking... but maybe this would need to be done generally for all passes in a follow-up.

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,315 @@
+    return num_failed;
+}
+
+int test_optimize_identity_target_inner(QkTarget *target) {
```
So far all functions called `test_..` are the main test entry points. I got a little confused when reading this as I expected the same. Could we rename this to something like `check_optimize_identity_with_target(QkTarget* target)` or something to clarify it's a helper?

raynelfss replied:
Addressed in 19f55a61197d9d6ac5b188778e4a2dd98b020b1c

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,315 @@
+    qk_transpiler_standalone_optimize_1q_gates_decomposition(circuit, target);
+    if (qk_circuit_count_ops(circuit).len != 0) {
+        result = EqualityError;
+        goto cleanup;
```
We can skip the goto statement since it is the next execution block anyways -- no need to proliferate this 😛 

Edit: the same holds also for some other tests here, we should only use the gotos if we really need them, they make the code harder to parse IMO

raynelfss replied:
Addressed in 19f55a61197d9d6ac5b188778e4a2dd98b020b1c

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,315 @@
+}
+
+/**
+ * "U is shorter than RZ-RY-RZ or RY-RZ-RY so use it when no error given.
```
```suggestion
 * U is shorter than RZ-RY-RZ or RY-RZ-RY so use it when no error given.
```

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,315 @@
+ *
+ * Transpile: 0:--[H]-[H]-[H]--
+ */
+int test_optimize_h_gates_inner(QkTarget *target, char **gates, uint32_t *freq, int num_gates) {
```
Same comment about the naming here 🙂

raynelfss replied:
Addressed in 19f55a61197d9d6ac5b188778e4a2dd98b020b1c

Cryoris reviewed (commented):
Overall LGTM, some nits below and then I think there are 2 open comments left: one about renaming this pass (which I still think we should take the chance for) and one about the docs saying that this pass compresses sequences into a single gate (which it doesn't, right?)

raynelfss committed:
Update test/c/test_optimize_1q_decomposition.c

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Chore: Address review comments.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-optimize-1q-decomp

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,310 @@
+        result = EqualityError;
+    }
+
+cleanup:
```
The cleanup label is no longer needed at all (same below in some other places 🙂)
```suggestion
```

raynelfss replied:
Removed in 923a8bb19018f55875382f5db28bc4d4c6dfb3e0

Cryoris commented:
This LGTM up to the two comments above on the documentation (this pass does not merge gates into a single gate, the Python docs are also wrong 🙂) and I still think we should rename this pass 🙂

raynelfss committed:
Fix: Review comments

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,307 @@
+
+    // Run transpiler pass
+    qk_transpiler_standalone_optimize_1q_sequences(circuit, target);
+    if (qk_circuit_count_ops(circuit).len != 0) {
```
Does the `QkOpCounts` object that is returned here get freed correctly? In the other C tests this is always done explicitly. Maybe we should explicitly allocate it and free it again? (Same for some other tests)

raynelfss replied:
Oh absolutely, I missed it here. I will add the `qk_ops_count_free()` calls.

raynelfss replied:
Added in 0a75d7c1ee32d8164cb5fe0f7f442b73393cf65e

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,307 @@
+
+    // Run transpiler pass
+    qk_transpiler_standalone_optimize_1q_sequences(circuit, get_rz_ry_u_noerror_target());
+    QkOpCounts counts = qk_circuit_count_ops(circuit);
```
Does this need to be freed via `qk_opcounts_free(counts)`?

raynelfss replied:
Freed in 0a75d7c1ee32d8164cb5fe0f7f442b73393cf65e

Cryoris commented on test/c/test_optimize_1q_decomposition.c:
```diff
@@ -0,0 +1,307 @@
+    qk_circuit_gate(circuit, QkGate_U, qubits, params);
+
+    // Run transpiler pass
+    qk_transpiler_standalone_optimize_1q_sequences(circuit, get_rz_ry_u_noerror_target());
```
It feels like we should properly free this `QkTarget*` again here, or don't we have to?

raynelfss replied:
Freed in 0a75d7c1ee32d8164cb5fe0f7f442b73393cf65e

Cryoris reviewed (commented):
some small comments about memory freeing otherwise I think we can merge 👍🏻

raynelfss committed:
Fix: Address some memory leakage.

raynelfss committed:
Fix: Incorrect function name call

raynelfss committed:
Chore: Address different review items
- Rename files from `optimize_1q_decomposition` to `optiize_1q_sequences`.
- Fixed previous oversight that counted operations before the pass was ran.

raynelfss committed:
Fix: Replace `QkOpCounts` failures until it is fixed.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-optimize-1q-decomp

Cryoris commented on test/c/test_optimize_1q_sequences.c:
```diff
@@ -0,0 +1,312 @@
+    return target_rz_ry_u_noerror;
+}
+
+bool compare_gate_counts(QkOpCounts counts, char **gates, uint32_t *freq, int num_gates) {
```
Since `QkOpCounts` is just queried here, I think we should take it by reference and then also call `free` in the caller and remove it from here 🙂

raynelfss replied:
Sounds like a cleaner approach.

raynelfss replied:
Added in 922425360ce157338604c8c84424fb7a83c6d22b

Cryoris commented on releasenotes/notes/c-api-optimize-1q-decomp-88954adfe952e91f.yaml:
```diff
@@ -0,0 +1,26 @@
+      qk_circuit_gate(circuit, QkGate_RY, qubits, params_pos);
+      qk_circuit_gate(circuit, QkGate_RY, qubits, params_neg);
+
+      // Run the transpiler pass with no Target
```
Could you clarify in this comment what happens if no target is supplied? 🙂

raynelfss replied:
Sure thing!

raynelfss replied:
Added in 922425360ce157338604c8c84424fb7a83c6d22b

Cryoris reviewed (dismissed):
LGTM, I have two optional comments which we can also clean up in a follow up.

raynelfss committed:
Fix: Add final review comments.
- Update release note to include explanation as to what happens if a `Target` is not provided.
- Update helper function `compare_gate_counts` to take `counts` by reference.

raynelfss committed:
Fix: Cformatting issues.

raynelfss committed:
Fix: More c formatting issues.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-optimize-1q-decomp
